rfcs/20231025-reusable-kernels/README.md

```diff
@@ -0,0 +1,106 @@
+kernels may depend on problem shape, but only in a limited way (i.e. the
+specific shape cannot be compiled directly into the kernel).
+
+## Proposal
+
+The proposal has four potential options: using DNNL_RUNTIME_DIM_VAL, a build
+flag, an environment variable, or a primitive attribute. In any of the options,
+the default behaviour will be the current one: kernels are ordered according to
+internal heuristics, roughly in order of decreasing likelihood for optimal
+performance for the given problem.
+
+If the reusable kernel option is selected, GPU kernels designed to be reusable
+will be selected first. In the case that no reusable kernels are compatible with
+the problem, kernel selection will continue with non-reusable kernels.
+
+Option 4 (primitive attribute) is the preferred option by frameworks, as they
+don't know which dimensions should be considered dynamic at runtime. It
+provides a balance between user control, flexibility, and ease of integration.
+Compared to the build flag it's more flexible, and compared to the environment
+variable it's less opaque.
+
+### Option 1: Using DNNL_RUNTIME_DIM_VAL
+
+With this option, the user will select one or more dimensions to have the
+DNNL_RUNTIME_DIM_VAL special value during primitive creation. oneDNN will then
+dispatch to the first implementation that supports the given set of runtime
+dimensions.
+<u>Pros:</u>
+- The most fine-grained approach
+- Ties into existing mechanisms
+- Can be toggled without a rebuild
+
+<u>Cons:</u>
+- Several primitives have no implementations that support runtime dims -
+cannot be adopted until such implementations have been developed
+- Support for all possible runtime dim configurations is difficult
+- Requires users to know which dimensions will have dynamic shapes ahead of
+time
+
+### Option 2: Build flag
+
+With this option, the build flag `ONEDNN_PREFER_REUSABLE_KERNELS` would be
+added to select the preferred functionality. Default is `off`, and reusable
+kernels will be chosen preferentially when set to `on`.
+<u>Pros:</u>
+- Easy integration
+
+<u>Cons:</u>
+- Requires a rebuild to change the option
+- The setting is not apparent in the code, leads to opaque behaviour
+- Requires internal changes to detect implementation reusability
+
+### Option 3: Environment variable
+
+With this option, the user can set `ONEDNN_PREFER_REUSABLE_KERNELS=1` as an
+environment variable, which will cause the library to select reusable GPU
+kernels preferentially. This option makes for easy integration, and provides
+a way to expand upon the functionality in the future (by setting a different
+value), for example if different levels of "reusability" are defined.
+<u>Pros:</u>
+- Easy integration
+- Allows for future expansion
+- Can be toggled without a rebuild
+
+<u>Cons:</u>
+- The setting is not apparent in the code, leads to opaque behaviour
+- Requires internal changes to detect implementation reusability
+
+### Option 4: Primitive attribute (preferred)
```

Comment:
Runtime dimensions was a mechanism we introduced to address the need for dynamic shapes and implemented for a subset of oneDNN functionality already. I would suggest we stick to it instead of introducing additional set of knobs serving similar purpose.

Reply:
Runtimes dimensions mechanism has two major issues (I think this may need to be better covered in the RFC):

- The mechanism is relatively hard for the frameworks to use (per their feedback) as it's not trivial to determine a subset of dimensions that are dynamic (for convolution: only spatial dimensions? filter sizes as well? what about strides?). It may be highly model-dependent which requires pre-analysis of the model on the framework side. It's not a simple solution.
- From the implementation side it's not trivial either. Even if we know some dimensions are dynamic - for decent performance we may need many kernels for good coverage. It's not clear when exactly to compile these different kernels. What if in the further runs only a subset of kernels will be used (then the user is paying extra overhead)? If we keep the implementation truly universal - performance is severely limited.

For the user it would be simpler to have mechanisms to control the reusability/performance trade-off at a high level depending on the scenario:
- Reusable kernels: fits dynamic size topologies to still be able to amortize primitive creation across multiple model iterations
- Full JIT: for the best performance when primitive creation time the model is fixed and there are many model iterations
- Auto-tuning/PGO: for the best performance when the user has enough time to pre-tune layers for a given model

Reply:
> For the user it would be simpler to have mechanisms to control the reusability/performance trade-off at a high level depending on the scenario

Actually, the feedback we got was that FWKs don't want to use such a mechanism since they don't have the information necessary: 
- in graph mode, only parameters have fixed shapes, but all other shapes can change from one graph execution to the next
- in eager mode, there is not even a notion of graph  or "unit" that is executed multiple times.

My understanding is that they would want primitive creation overheads hidden under oneDNN transparently.

> From the implementation side it's not trivial either. Even if we know some dimensions are dynamic - for decent performance we may need many kernels for good coverage. It's not clear when exactly to compile these different kernels. What if in the further runs only a subset of kernels will be used (then the user is paying extra overhead)? If we keep the implementation truly universal - performance is severely limited.

I think we are overthinking here. There are only a few actual CNN use-cases in practice:
- all shapes are static
- only minibatch changes (some inference scenarios)
- both minibatch and spatials change from one execution to the next (object detection)
I guess a good starting point would be to cover the first and last bullets (which covers the second bullet).

Reply:
> My understanding is that they would want primitive creation overheads hidden under oneDNN transparently.

Unfortunately this is not feasible in the short term hence we came up with this RFC. Ideally we want fast kernels with no overhead on creation - this would be the perfect scenario but it's not realistic - at least in the near future.


> I think we are overthinking here. There are only a few actual CNN use-cases in practice:

> all shapes are static
only minibatch changes (some inference scenarios)
both minibatch and spatials change from one execution to the next (object detection)
I guess a good starting point would be to cover the first and last bullets (which covers the second bullet).

Right but I still don't see how to address the issues I mentioned - dynamic batch/spatial may require several kernels (or dozens of kernels for JIT-level performance) - it's not practical to compile them all when say in practice only 1/10 of all kernels are used.

Reply:
> Right but I still don't see how to address the issues I mentioned - dynamic batch/spatial may require several kernels (or dozens of kernels for JIT-level performance) - it's not practical to compile them all when say in practice only 1/10 of all kernels are used.

Correct, this is still part of the creation time vs execution time tradeoff. How does the hint helps though?

Reply:
The hint helps to amortize creation time via reusable kernels. The end goal is the same as with `DNNL_RUNTIME_DIM_VAL` feature while the hint doesn't provide any strict guarantees.

The main problem with `DNNL_RUNTIME_DIM_VAL` is the API imposed flow when the user is asked to create a primitive once to be able to call it for any sizes. With such strict guarantees we can't provide a reasonable creation time and good enough performance - the set of sizes to cover is too large. With the hint it's relaxed - we generate kernels on demand without strict guarantees.

Reply:
ok I think I see now thanks! So in a nutshell:
- using DNNL_RUNTIME_DIM_VAL: implementation X which handle dynamic shapes has to jit all its kernels upon creation
- using hint:  implementation X which handle dynamic shapes jit only the subset of kernels for the passed shape

So in first case, all is jitted upon first iteration, and then no more jitting.
For the second case, some jitting happen upon first iteration, and extra jitting might or might not happen depending on the useful subset of kernels already jitted.

I guess two things here:
- do we have data on how big the gap is between the two options?
- what about AOT compilation?

Reply:
> do we have data on how big the gap is between the two options?

This is in development, and it will take some time until we can say that our "reusable" kernels have all important optimizations for such a comparison to make sense. As a side point, we are quite sure that the JIT option can always provide an extra boost (say at least 5-10%) so it's probably a good idea for oneDNN to allow users to switch between JIT/reusable kernels depending on the scenario.

> what about AOT compilation?

This is an option but it needs more exploring. At a first glance, I see two major arguments against AOT:

- It may not be feasible - consider all data type and post-op combinations. If covering everything the library size explodes significantly. It might be we can only do limited AOT support
- I don't think we have any known scenarios when AOT is practically needed. I expect for real model runs we have enough iterations to amortize JIT compilation time. However this is only possible with reusable kernels - when kernels are reused between iterations even for dynamic shapes.